proposals/progress_output.md

```diff
@@ -0,0 +1,177 @@
+
+This is especially prevalent for first-time users whose first builds would need to push all the layers, including the large base image layers.
+
+## Current output
+
+An example of the current output (as of versoin `0.10.0`) looks like:
+
+```
+Containerizing application to <image>...
+warning: Base image 'gcr.io/distroless/java' does not use a specific image digest - build may not be reproducible
+Retrieving registry credentials for gcr.io...
+Getting base image gcr.io/distroless/java...
+Building dependencies layer...
+Building resources layer...
+Building classes layer...
+
+Container entrypoint set to [java, -cp, /app/resources:/app/classes:/app/libs/*, <main class>]
+Finalizing...
+
+Built and pushed image as <image>
+```
+
+Note that the build would pause on `Finalizing...` until all layers are finished pushing.
+
+## Goal
+
+To display fine-grained progress feedback during the build, but not clutter the log messages.
+
+## Design considerations
+
+Since Jib builds/pulls and pushes each layer independently, synchronized progress feedback (completion over all layers, such as a progress bar) could potentially limit the concurrency of the build process - but this might be negligible.
+
+The available information for each layer include:
+
+- type (classes, resources, dependencies, base image layer, etc.)
+- digest
+- size
+
+The non-available information for each layer include:
+
+- sizes of other layers (eg. other layers may not have finished building)
+- number of other layers (eg. base image layers have not been resolved before starting to push an application layer)
+
+Therefore, the actions we can perform for progress feedback include:
+
+- monitor the number of bytes sent (and percentage completion)
+
+And the actions we cannot perform include:
+
+- know ahead of time how many layers will be pulled/pushed
+
+## Proposal
+
+Display an overall progress bar along with the tasks currently being executed.
+
+### Example
+
+```
+Executing tasks: Pushing classes layer, pulling base image layer 50501d3b88f7, pushing dependencies layer, pushing base image layer 8b106a18283f
+[========================         ] 80% complete
+```
+
+*Note that this would replace the logs messages that are currently outputted as those may corrupt the progress bar display.*
+
+### Summary
+
+The implementation consists of two parts:
+
+1. Emit progress events from the builder steps.
+1. Monitor progress events and display to console.
+
+### Emit progress events
+
+There are a few issues to address for the progress event design:
+
+- the total amount of work is not known beforehand, so there is no static max progress known at any time
+- the builder steps are asynchronous and therefore progress events should not share a single progress state
+- rather, the progress event receiver should be the only potentially stateful entity
+
+These issues can be resolved with a *decentralized allocation tree*.
+
+#### Decentralized allocation tree (DAT)
```

Comment:
I read through it, and still not clear if this will work as intended, if what I understood is correct, which makes me think probably I'm missing something.

So, what I can only think of for the intention of a tree is that, basically, the root node corresponds to the top-level entity that kicks off multiple second-level async "steps". For example, if you look at `BuildSteps.forBuildToDockerRegistry`, it spawns 12 steps:

```java
  public static BuildSteps forBuildToDockerRegistry(BuildConfiguration buildConfiguration) {
    return new BuildSteps(
        DESCRIPTION_FOR_DOCKER_REGISTRY,
        buildConfiguration,
        () ->
            new StepsRunner(buildConfiguration)
                .runRetrieveTargetRegistryCredentialsStep()
                .runAuthenticatePushStep()
                .runPullBaseImageStep()
                .runPullAndCacheBaseImageLayersStep()
                .runPushBaseImageLayersStep()
                .runBuildAndCacheApplicationLayerSteps()
                .runBuildImageStep()
                .runPushContainerConfigurationStep()
                .runPushApplicationLayersStep()
                .runFinalizingPushStep()
                .runPushImageStep()
                .waitOnPushImageStep());
  }
```
So you intend that you create a root node here with 12 allocation units, pass the root node to the 12 `Step` instances, and each step will create a single child node (with some allocation units) soon, resulting in 13 nodes (1 root + 12 children) at some point? And each `Step` passes down its child node down to methods and instances they own in the similar manner?

Reply:
Yes, that is how it is intended to work. The only thing that will need to known prior to an allocation node being created is the number of allocation units that node should have.

Reply:
Oh one thing to note is that the root node with 12 allocation units does not necessarily need to have 12 children - allocation units can be occupied by non-node-based updates (like received bytes in the download progress case).

Reply:
If so (passing down the root node to multiple async step instances), and if the root and the children are physically connected (I assume so, because otherwise, it isn't a tree at all), isn't it that you are sharing the tree (at least the root node) concurrently? It sounds like basically you are sharing a single progress state. I thought it'd be cool if anyone could just fire progress events in an isolated manner, not having to establish any connection in relation to others. With no communications with other, it may be impossible to avoid the "backward-moving" progress bar, but I think it's worth the sacrifice for simplicity.

Reply:
The node(s) are passed down, but they are all immutable with no progress state. The state is only kept at the receiving end (the progress monitor that receives the progress events). The DAT is essentially just a representation of what fraction of the total progress a progress event actually accounts for. This allows for progress events to be emitted asynchronously and in a decentralized manner, only to be collected by the progress monitor and summed into an overall current progress.